pkg/fqdn/cache.go

```diff
@@ -154,16 +178,125 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
 		c.forward[entry.Name] = entries
 	}
 	c.updateWithEntryIPs(entries, entry)
+	if len(entries) > c.perHostLimit {
```

Comment:
Oh, also, this should go after c.removeExpired below, since after expiration we may not be over the limit.
Similarly, we should check if we are under the limit, and remove ourselves from overlimit:
```
switch {
  case c.perHostLimit > 0 && len(entries) > c.perHostLimit:
    c.overLimit[entry.Name] = true
  case c.perHostLimit > 0 && len(entries) < c.perHostLimit:
     delete(c.overLimit, entry.Name)
}

Reply:
Good one, the delete does not matter here. We already checked on the GC function ;-)